[PyTorch] Stack-allocate boxed args for RecordFunction

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #76266

Saving a heap allocation in this path improves performance.

Differential Revision: [D34090699](https://our.internmc.facebook.com/intern/diff/D34090699/)

swolchok committed:
[PyTorch] Stack-allocate boxed args for RecordFunction

Saving a heap allocation in this path improves performance.

Differential Revision: [D34090699](https://our.internmc.facebook.com/intern/diff/D34090699/)

[ghstack-poisoned]

swolchok commented:
An interesting follow-up would be to make it possible to avoid heap-allocating a Stack for calls to boxed KernelFunctions made via make_boxed_from_unboxed_functor. I have no idea how common this is, though.

ezyang commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -490,6 +490,52 @@ struct CaptureKernelCall<void> {
+template <typename T>
+C10_DISPATCHER_INLINE_UNLESS_MOBILE void box(IValueStorage* dest, T& arg, int& lastIdx) {
+  new (&dest[lastIdx]) IValue(arg);
+  lastIdx++;
```
This needs to be kept consistent with our other boxing logic, is that right?

ezyang replied:
I'd suggest a cross-ref with aten/src/ATen/core/boxing/impl/boxing.h (or perhaps moving it to this file entirely)

ezyang commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -490,6 +490,52 @@ struct CaptureKernelCall<void> {
+using IValueStorage = std::aligned_storage_t<sizeof(IValue), alignof(IValue)>;
+
+template <typename T>
+C10_DISPATCHER_INLINE_UNLESS_MOBILE void box(IValueStorage* dest, T& arg, int& lastIdx) {
```
The name here is a bit ambiguous; what we're doing is placement new boxing; would be nice if the name had this so people don't try to use this invalidly.

swolchok replied:
it's in detail and takes a weird typedef; they'll have trouble using it without writing obvious "here be dragons" stuff like aligned_storage

ezyang reviewed (approved):
mechanically looks all fine, just some naming / file placement stuff

swolchok committed:
Update on "[PyTorch] Stack-allocate boxed args for RecordFunction"


Saving a heap allocation in this path improves performance.

Differential Revision: [D34090699](https://our.internmc.facebook.com/intern/diff/D34090699/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Stack-allocate boxed args for RecordFunction"


Saving a heap allocation in this path improves performance.

Differential Revision: [D34090699](https://our.internmc.facebook.com/intern/diff/D34090699/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Stack-allocate boxed args for RecordFunction"


Saving a heap allocation in this path improves performance.

Differential Revision: [D34090699](https://our.internmc.facebook.com/intern/diff/D34090699/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Stack-allocate boxed args for RecordFunction"


Saving a heap allocation in this path improves performance.

Differential Revision: [D34090699](https://our.internmc.facebook.com/intern/diff/D34090699/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Stack-allocate boxed args for RecordFunction"


Saving a heap allocation in this path improves performance.

Differential Revision: [D34090699](https://our.internmc.facebook.com/intern/diff/D34090699/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Stack-allocate boxed args for RecordFunction"


Saving a heap allocation in this path improves performance.

Differential Revision: [D34090699](https://our.internmc.facebook.com/intern/diff/D34090699/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Stack-allocate boxed args for RecordFunction"


Saving a heap allocation in this path improves performance.

Differential Revision: [D34090699](https://our.internmc.facebook.com/intern/diff/D34090699/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Stack-allocate boxed args for RecordFunction"


Saving a heap allocation in this path improves performance.

Differential Revision: [D34090699](https://our.internmc.facebook.com/intern/diff/D34090699/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Stack-allocate boxed args for RecordFunction"


Saving a heap allocation in this path improves performance.

Differential Revision: [D34090699](https://our.internmc.facebook.com/intern/diff/D34090699/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Stack-allocate boxed args for RecordFunction"


Saving a heap allocation in this path improves performance.

Differential Revision: [D34090699](https://our.internmc.facebook.com/intern/diff/D34090699/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Stack-allocate boxed args for RecordFunction"


Saving a heap allocation in this path improves performance.

Differential Revision: [D34090699](https://our.internmc.facebook.com/intern/diff/D34090699/)

[ghstack-poisoned]
